From e5e5ed801e612caf193115e800fb293fce9c22b2 Mon Sep 17 00:00:00 2001
From: Yannick Lanz <yannick.lanz@wifx.net>
Date: Thu, 5 Dec 2024 22:23:11 +0100
Subject: [PATCH 3/3] vendor: Wifx LORIX One and L1 support, support for 8XX
 and 9XX LoRa variant with frontend revision. Only for testing, revision B is
 a copy of revision A for the moment.

Signed-off-by: Yannick Lanz <yannick.lanz@wifx.net>
---
 .../src/config/mod.rs                         |   6 +-
 ...orix_one_863_870.rs => lorix_one_8XX_A.rs} |   0
 ...orix_one_902_928.rs => lorix_one_9XX_A.rs} |   0
 .../src/config/vendor/wifx/lorix_one_eu868.rs | 288 ------------------
 .../src/config/vendor/wifx/mod.rs             |   6 +-
 .../src/config/mod.rs                         |   6 +-
 .../wifx/{l1_863_870.rs => l1_8XX_A.rs}       |   0
 .../src/config/vendor/wifx/l1_8XX_B.rs        | 103 +++++++
 .../wifx/{l1_902_928.rs => l1_9XX_A.rs}       |   0
 .../src/config/vendor/wifx/l1_9XX_B.rs        | 103 +++++++
 .../src/config/vendor/wifx/mod.rs             |  10 +-
 11 files changed, 224 insertions(+), 298 deletions(-)
 rename chirpstack-concentratord-sx1301/src/config/vendor/wifx/{lorix_one_863_870.rs => lorix_one_8XX_A.rs} (100%)
 rename chirpstack-concentratord-sx1301/src/config/vendor/wifx/{lorix_one_902_928.rs => lorix_one_9XX_A.rs} (100%)
 delete mode 100644 chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_eu868.rs
 rename chirpstack-concentratord-sx1302/src/config/vendor/wifx/{l1_863_870.rs => l1_8XX_A.rs} (100%)
 create mode 100644 chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_8XX_B.rs
 rename chirpstack-concentratord-sx1302/src/config/vendor/wifx/{l1_902_928.rs => l1_9XX_A.rs} (100%)
 create mode 100644 chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_9XX_B.rs

diff --git a/chirpstack-concentratord-sx1301/src/config/mod.rs b/chirpstack-concentratord-sx1301/src/config/mod.rs
index eebf67d..855f9da 100644
--- a/chirpstack-concentratord-sx1301/src/config/mod.rs
+++ b/chirpstack-concentratord-sx1301/src/config/mod.rs
@@ -202,10 +202,8 @@ pub fn get(filenames: Vec<String>) -> Configuration {
         "risinghf_rhf0m301_us915" => vendor::risinghf::rhf0m301_us915::new(&config),
         "sandbox_lorago_port_eu868" => vendor::sandbox::lorago_port_eu868::new(&config),
         "sandbox_lorago_port_us915" => vendor::sandbox::lorago_port_us915::new(&config),
-        "wifx_lorix_one_863_870" | "wifx_lorix_one_eu868" => {
-            vendor::wifx::lorix_one_863_870::new(&config)
-        }
-        "wifx_lorix_one_902_928" => vendor::wifx::lorix_one_902_928::new(&config),
+        "wifx_lorix_one_8XX_A" => vendor::wifx::lorix_one_8XX_A::new(&config),
+        "wifx_lorix_one_9XX_A" => vendor::wifx::lorix_one_9XX_A::new(&config),
         _ => panic!("unexpected gateway model: {}", config.gateway.model),
     };
 
diff --git a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_863_870.rs b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_8XX_A.rs
similarity index 100%
rename from chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_863_870.rs
rename to chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_8XX_A.rs
diff --git a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_902_928.rs b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_9XX_A.rs
similarity index 100%
rename from chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_902_928.rs
rename to chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_9XX_A.rs
diff --git a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_eu868.rs b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_eu868.rs
deleted file mode 100644
index 19fa9c4..0000000
--- a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_eu868.rs
+++ /dev/null
@@ -1,288 +0,0 @@
-use libloragw_sx1301::hal;
-
-use super::super::super::super::config;
-use super::super::Configuration;
-
-// source:
-// https://github.com/Wifx/meta-wifx/blob/krogoth/recipes-connectivity/packet-forwarder/files/configs/global_conf_EU868_2dBi_indoor.json
-// https://github.com/Wifx/meta-wifx/blob/krogoth/recipes-connectivity/packet-forwarder/files/configs/global_conf_EU868_4dBi_outdoor.json
-pub fn new(conf: &config::Configuration) -> Configuration {
-    Configuration {
-        radio_count: 2,
-        clock_source: 1,
-        radio_rssi_offset: vec![-164.0, -164.0],
-        radio_tx_enabled: vec![true, false],
-        radio_type: vec![hal::RadioType::SX1257, hal::RadioType::SX1257],
-        radio_min_max_tx_freq: vec![(863000000, 870000000), (863000000, 870000000)],
-        radio_tx_notch_freq: vec![129000, 0],
-        lora_multi_sf_bandwidth: 125000,
-        tx_gain_table: if conf.gateway.antenna_gain == 2 {
-            vec![
-                // 0
-                hal::TxGainConfig {
-                    pa_gain: 0,
-                    mix_gain: 10,
-                    rf_power: -6,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 1
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 9,
-                    rf_power: -3,
-                    dig_gain: 3,
-                    dac_gain: 3,
-                },
-                // 2
-                hal::TxGainConfig {
-                    pa_gain: 0,
-                    mix_gain: 15,
-                    rf_power: 0,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 3
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 9,
-                    rf_power: 3,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 4
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 8,
-                    rf_power: 6,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 5
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 10,
-                    rf_power: 10,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 6
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 9,
-                    rf_power: 11,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 7
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 11,
-                    rf_power: 12,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 8
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 11,
-                    rf_power: 13,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 9
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 12,
-                    rf_power: 14,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 10
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 12,
-                    rf_power: 16,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 11
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 12,
-                    rf_power: 20,
-                    dig_gain: 3,
-                    dac_gain: 3,
-                },
-                // 12
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 11,
-                    rf_power: 23,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 13
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 12,
-                    rf_power: 25,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 14
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 12,
-                    rf_power: 26,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 15
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 13,
-                    rf_power: 27,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-            ]
-        } else if conf.gateway.antenna_gain == 4 {
-            vec![
-                // 0
-                hal::TxGainConfig {
-                    pa_gain: 0,
-                    mix_gain: 9,
-                    rf_power: -6,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 1
-                hal::TxGainConfig {
-                    pa_gain: 0,
-                    mix_gain: 10,
-                    rf_power: -3,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 2
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 8,
-                    rf_power: 0,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 3
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 11,
-                    rf_power: 3,
-                    dig_gain: 3,
-                    dac_gain: 3,
-                },
-                // 4
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 13,
-                    rf_power: 6,
-                    dig_gain: 3,
-                    dac_gain: 3,
-                },
-                // 5
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 13,
-                    rf_power: 10,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 6
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 15,
-                    rf_power: 11,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 7
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 10,
-                    rf_power: 12,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 8
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 10,
-                    rf_power: 13,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 9
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 11,
-                    rf_power: 14,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 10
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 11,
-                    rf_power: 16,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 11
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 15,
-                    rf_power: 20,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 12
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 10,
-                    rf_power: 23,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 13
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 11,
-                    rf_power: 25,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 14
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 11,
-                    rf_power: 26,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 15
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 15,
-                    rf_power: 27,
-                    dig_gain: 3,
-                    dac_gain: 3,
-                },
-            ]
-        } else {
-            panic!("Invalid antenna_gain: {}", conf.gateway.antenna_gain);
-        },
-        gps_tty_path: None,
-        spidev_path: "/dev/spidev0.0".to_string(),
-        reset_pin: Some((0, 1)),
-    }
-}
diff --git a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/mod.rs b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/mod.rs
index 4fdecef..9256aee 100644
--- a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/mod.rs
+++ b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/mod.rs
@@ -1,2 +1,4 @@
-pub mod lorix_one_863_870;
-pub mod lorix_one_902_928;
+#[allow(non_snake_case)]
+pub mod lorix_one_8XX_A;
+#[allow(non_snake_case)]
+pub mod lorix_one_9XX_A;
diff --git a/chirpstack-concentratord-sx1302/src/config/mod.rs b/chirpstack-concentratord-sx1302/src/config/mod.rs
index b2f41c7..4bb0d32 100644
--- a/chirpstack-concentratord-sx1302/src/config/mod.rs
+++ b/chirpstack-concentratord-sx1302/src/config/mod.rs
@@ -172,8 +172,10 @@ pub fn get(filenames: Vec<String>) -> Configuration {
         "rak_2287_kr920" => vendor::rak::rak2287_kr920::new(&config),
         "rak_2287_ru864" => vendor::rak::rak2287_ru864::new(&config),
         "rak_2287_us915" => vendor::rak::rak2287_us915::new(&config),
-        "wifx_l1_863_870" => vendor::wifx::l1_863_870::new(&config),
-        "wifx_l1_902_928" => vendor::wifx::l1_902_928::new(&config),
+        "wifx_l1_8XX_A" => vendor::wifx::l1_8XX_A::new(&config),
+        "wifx_l1_8XX_B" => vendor::wifx::l1_8XX_B::new(&config),
+        "wifx_l1_9XX_A" => vendor::wifx::l1_9XX_A::new(&config),
+        "wifx_l1_9XX_B" => vendor::wifx::l1_9XX_B::new(&config),
         _ => panic!("unexpected gateway model: {}", config.gateway.model),
     };
 
diff --git a/chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_863_870.rs b/chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_8XX_A.rs
similarity index 100%
rename from chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_863_870.rs
rename to chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_8XX_A.rs
diff --git a/chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_8XX_B.rs b/chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_8XX_B.rs
new file mode 100644
index 0000000..aef2c13
--- /dev/null
+++ b/chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_8XX_B.rs
@@ -0,0 +1,103 @@
+use libloragw_sx1302::hal;
+
+use super::super::super::super::config;
+use super::super::{ComType, Configuration, RadioConfig};
+
+pub fn new(conf: &config::Configuration) -> Configuration {
+    Configuration {
+        radio_count: 2,
+        clock_source: 1,
+        full_duplex: false,
+        lora_multi_sf_bandwidth: 125000,
+        radio_config: vec![
+            RadioConfig {
+                enable: true,
+                radio_type: hal::RadioType::SX1250,
+                single_input_mode: true,
+                rssi_offset: -215.4,
+                rssi_temp_compensation: hal::RssiTempCompensationConfig {
+                    coeff_a: 0.0,
+                    coeff_b: 0.0,
+                    coeff_c: 20.41,
+                    coeff_d: 2162.56,
+                    coeff_e: 0.0,
+                },
+                tx_enable: true,
+                tx_freq_min: 863000000,
+                tx_freq_max: 870000000,
+                tx_gain_table: super::get_tx_gain_table(
+                    POWER_TABLE_BASE.to_vec(),
+                    conf.gateway.antenna_gain,
+                ),
+            },
+            RadioConfig {
+                enable: true,
+                radio_type: hal::RadioType::SX1250,
+                single_input_mode: false,
+                rssi_offset: -215.4,
+                rssi_temp_compensation: hal::RssiTempCompensationConfig {
+                    coeff_a: 0.0,
+                    coeff_b: 0.0,
+                    coeff_c: 20.41,
+                    coeff_d: 2162.56,
+                    coeff_e: 0.0,
+                },
+                tx_enable: false,
+                tx_freq_min: 0,
+                tx_freq_max: 0,
+                tx_gain_table: vec![],
+            },
+        ],
+        gps_tty_path: None,
+        com_type: ComType::SPI,
+        com_path: "/dev/spidev0.0".to_string(),
+        reset_pin: match conf.gateway.reset_pin {
+            0 => Some((0, 1)),
+            _ => Some((0, conf.gateway.reset_pin)),
+        },
+        power_en_pin: match conf.gateway.power_en_pin {
+            0 => None,
+            _ => Some((0, conf.gateway.power_en_pin)),
+        },
+    }
+}
+
+// Power table with an antenna gain of 0dBi
+lazy_static! {
+    static ref POWER_TABLE_BASE: [hal::TxGainConfig; 21] = [
+        gain_param(7, 0, 7),
+        gain_param(8, 0, 8),
+        gain_param(9, 0, 9),
+        gain_param(10, 0, 10),
+        gain_param(11, 0, 11),
+        gain_param(12, 0, 12),
+        gain_param(13, 0, 13),
+        gain_param(14, 0, 14),
+        gain_param(15, 0, 15),
+        gain_param(16, 0, 16),
+        gain_param(17, 0, 18),
+        gain_param(18, 0, 19),
+        gain_param(19, 1, 0),
+        gain_param(20, 1, 1),
+        gain_param(21, 1, 2),
+        gain_param(22, 1, 3),
+        gain_param(23, 1, 4),
+        gain_param(24, 1, 5),
+        gain_param(25, 1, 7),
+        gain_param(26, 1, 9),
+        gain_param(27, 1, 14),
+    ];
+}
+
+fn gain_param(rf_power: i8, pa: u8, pwr_idx: u8) -> hal::TxGainConfig {
+    hal::TxGainConfig {
+        rf_power: rf_power,
+        dig_gain: 0,
+        pa_gain: pa,
+        dac_gain: 0,
+        mix_gain: 5,
+        offset_i: 0,
+        offset_q: 0,
+        pwr_idx: pwr_idx,
+    }
+}
diff --git a/chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_902_928.rs b/chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_9XX_A.rs
similarity index 100%
rename from chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_902_928.rs
rename to chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_9XX_A.rs
diff --git a/chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_9XX_B.rs b/chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_9XX_B.rs
new file mode 100644
index 0000000..7526834
--- /dev/null
+++ b/chirpstack-concentratord-sx1302/src/config/vendor/wifx/l1_9XX_B.rs
@@ -0,0 +1,103 @@
+use libloragw_sx1302::hal;
+
+use super::super::super::super::config;
+use super::super::{ComType, Configuration, RadioConfig};
+
+pub fn new(conf: &config::Configuration) -> Configuration {
+    Configuration {
+        radio_count: 2,
+        clock_source: 1,
+        full_duplex: false,
+        lora_multi_sf_bandwidth: 125000,
+        radio_config: vec![
+            RadioConfig {
+                enable: true,
+                radio_type: hal::RadioType::SX1250,
+                single_input_mode: true,
+                rssi_offset: -215.4,
+                rssi_temp_compensation: hal::RssiTempCompensationConfig {
+                    coeff_a: 0.0,
+                    coeff_b: 0.0,
+                    coeff_c: 20.41,
+                    coeff_d: 2162.56,
+                    coeff_e: 0.0,
+                },
+                tx_enable: true,
+                tx_freq_min: 902000000,
+                tx_freq_max: 928000000,
+                tx_gain_table: super::get_tx_gain_table(
+                    POWER_TABLE_BASE.to_vec(),
+                    conf.gateway.antenna_gain,
+                ),
+            },
+            RadioConfig {
+                enable: true,
+                radio_type: hal::RadioType::SX1250,
+                single_input_mode: false,
+                rssi_offset: -215.4,
+                rssi_temp_compensation: hal::RssiTempCompensationConfig {
+                    coeff_a: 0.0,
+                    coeff_b: 0.0,
+                    coeff_c: 20.41,
+                    coeff_d: 2162.56,
+                    coeff_e: 0.0,
+                },
+                tx_enable: false,
+                tx_freq_min: 0,
+                tx_freq_max: 0,
+                tx_gain_table: vec![],
+            },
+        ],
+        gps_tty_path: None,
+        com_type: ComType::SPI,
+        com_path: "/dev/spidev0.0".to_string(),
+        reset_pin: match conf.gateway.reset_pin {
+            0 => Some((0, 1)),
+            _ => Some((0, conf.gateway.reset_pin)),
+        },
+        power_en_pin: match conf.gateway.power_en_pin {
+            0 => None,
+            _ => Some((0, conf.gateway.power_en_pin)),
+        },
+    }
+}
+
+// Power table with an antenna gain of 0dBi
+lazy_static! {
+    static ref POWER_TABLE_BASE: [hal::TxGainConfig; 21] = [
+        gain_param(7, 0, 8),
+        gain_param(8, 0, 9),
+        gain_param(9, 0, 10),
+        gain_param(10, 0, 11),
+        gain_param(11, 0, 12),
+        gain_param(12, 0, 13),
+        gain_param(13, 0, 14),
+        gain_param(14, 0, 15),
+        gain_param(15, 0, 16),
+        gain_param(16, 0, 17),
+        gain_param(17, 0, 18),
+        gain_param(18, 0, 19),
+        gain_param(19, 1, 0),
+        gain_param(20, 0, 22),
+        gain_param(21, 1, 2),
+        gain_param(22, 1, 3),
+        gain_param(23, 1, 4),
+        gain_param(24, 1, 6),
+        gain_param(25, 1, 8),
+        gain_param(26, 1, 11),
+        gain_param(27, 1, 15),
+    ];
+}
+
+fn gain_param(rf_power: i8, pa: u8, pwr_idx: u8) -> hal::TxGainConfig {
+    hal::TxGainConfig {
+        rf_power: rf_power,
+        dig_gain: 0,
+        pa_gain: pa,
+        dac_gain: 0,
+        mix_gain: 5,
+        offset_i: 0,
+        offset_q: 0,
+        pwr_idx: pwr_idx,
+    }
+}
diff --git a/chirpstack-concentratord-sx1302/src/config/vendor/wifx/mod.rs b/chirpstack-concentratord-sx1302/src/config/vendor/wifx/mod.rs
index b0eb8f6..5138d4a 100644
--- a/chirpstack-concentratord-sx1302/src/config/vendor/wifx/mod.rs
+++ b/chirpstack-concentratord-sx1302/src/config/vendor/wifx/mod.rs
@@ -1,5 +1,11 @@
-pub mod l1_863_870;
-pub mod l1_902_928;
+#[allow(non_snake_case)]
+pub mod l1_8XX_A;
+#[allow(non_snake_case)]
+pub mod l1_8XX_B;
+#[allow(non_snake_case)]
+pub mod l1_9XX_A;
+#[allow(non_snake_case)]
+pub mod l1_9XX_B;
 
 use libloragw_sx1302::hal;
 
-- 
2.34.1

